src/server/scripts/EasternKingdoms/zone_isle_of_queldanas.cpp

```diff
@@ -0,0 +1,429 @@
+        void DoAction(int32 action) override
+        {
+            if (action == ACTION_START_QUEST)
+            {
+                me->RemoveFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_GOSSIP);
+
+                if (Creature* thalorien = me->SummonCreature(NPC_THALORIEN_DAWNSEEKER, thalorienSummon, TEMPSUMMON_CORPSE_DESPAWN, 0))
+                    _thalorienGUID = thalorien->GetGUID();
+                me->SummonCreatureGroup(SUMMON_SUNWELL_DEFENDER);
+
+                _events.ScheduleEvent(EVENT_INTRO_0, 3s);
+            }
+        }
+
+        void Reset() override
+        {
+            me->SetFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_GOSSIP);
+            _summons.DespawnAll();
+            _playerGUID.Clear();
+            _morlenGUID.Clear();
+            _thalorienGUID.Clear();
+            _events.Reset();
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            _events.Update(diff);
+
+            while (uint32 eventId = _events.ExecuteEvent())
+            {
+                switch (eventId)
+                {
+                    case EVENT_INTRO_0:
+                        if (Creature* thalorien = ObjectAccessor::GetCreature(*me, _thalorienGUID))
+                            thalorien->AI()->Talk(SAY_THALORIEN_1);
+
+                        _events.ScheduleEvent(EVENT_INTRO_1, 5s);
+                        break;
+                    case EVENT_INTRO_1:
+                        if (Creature* thalorien = ObjectAccessor::GetCreature(*me, _thalorienGUID))
+                            thalorien->AI()->Talk(SAY_THALORIEN_2);
+
+                        _events.ScheduleEvent(EVENT_INTRO_2, 6s);
+                        break;
+                    case EVENT_INTRO_2:
+                        if (Creature* thalorien = ObjectAccessor::GetCreature(*me, _thalorienGUID))
+                            thalorien->AI()->Talk(SAY_THALORIEN_3);
+
+                        _events.ScheduleEvent(EVENT_INTRO_3, 7s);
+                        break;
+
+                    case EVENT_INTRO_3:
+                        if (Creature* thalorien = ObjectAccessor::GetCreature(*me, _thalorienGUID))
+                            thalorien->AI()->Talk(SAY_THALORIEN_4);
+
+                        _events.ScheduleEvent(EVENT_SALUTE, 4s);
+                        break;
+                    case EVENT_SALUTE:
+                        for (auto& summon : _summons)
+                            if (Creature* creature = ObjectAccessor::GetCreature(*me, summon))
+                                creature->HandleEmoteCommand(EMOTE_ONESHOT_SALUTE);
+
+                        _events.ScheduleEvent(EVENT_DEFENDERS_RUN, 3s);
+                        break;
+                    case EVENT_DEFENDERS_RUN:
+                    {
+                        _events.ScheduleEvent(EVENT_DEFENDERS_RUN_2, 3s);
+                        _events.ScheduleEvent(EVENT_THALORIEN_GO, 2s);
+
+                        uint8 i = 0;
+                        for (auto& summon : _summons)
```

Comment:
Why range based loop over standard loop if you're iterating a value anyway?

Reply:
I can access list objects easily using `auto&`.